### PR TITLE
chore(deps): update dependency rimraf to v5.0.10

### LINKED ISSUE #30148: Update to `rimraf` 4.0.0 or later
### What would you like?

1. Update all uses of the npm module [rimraf](https://www.npmjs.com/package/rimraf) to minimum [rimraf@4.0.0](https://github.com/isaacs/rimraf/tree/v4.0.0), preferably [rimraf@5.0.10](https://github.com/isaacs/rimraf/tree/v5.0.10) (or latest `5.x`). This is the highest version which is still compatible with Node.js `18.x` (see [CHANGELOG](https://github.com/isaacs/rimraf/blob/main/CHANGELOG.md)).
2. Remove the direct use of the deprecated npm module [@types/rimraf](https://www.npmjs.com/package/@types/rimraf).

### Why is this needed?

The npm module [@types/rimraf](https://www.npmjs.com/package/@types/rimraf) is deprecated. [rimraf@4.0.0](https://github.com/isaacs/rimraf/tree/v4.0.0) introduced TypeScript support - also mentioned in the [CHANGELOG v4.0](https://github.com/isaacs/rimraf/blob/main/CHANGELOG.md#v40).

### Other

- Renovate deprecation list https://github.com/cypress-io/cypress/issues/3777

| Item                      | Value                                                                                                         |
| ------------------------- | ------------------------------------------------------------------------------------------------------------- |
| Deprecated npm module     | [@types/rimraf](https://www.npmjs.com/package/@types/rimraf)                                                  |
| npm registry message      | This is a stub types definition. rimraf provides its own type definitions, so you do not need this installed. |
| Other message             |                                                                                                               |
| GitHub repo               | https://github.com/isaacs/rimraf                                                                              |
| Repo status               | Active                                                                                                        |
| Replacement npm module    | https://www.npmjs.com/package/rimraf                                                                          |
| Replacement repo          | https://github.com/isaacs/rimraf                                                                              |
| Replacement documentation | https://github.com/isaacs/rimraf                                                                              |
| Migration docs            | https://github.com/isaacs/rimraf                                                                              |

Where used:

| Directory                                                                             | Version                                            |
| ------------------------------------------------------------------------------------- | -------------------------------------------------- |
| [repo root](https://github.com/cypress-io/cypress)                                    | `"@types/rimraf": "^3.0.2"`<br>`"rimraf": "3.0.2"` |
| [npm/puppeteer](https://github.com/cypress-io/cypress/tree/develop/npm/puppeteer)     | `"rimraf": "^5.0.1"`                               |
| [packages/errors](https://github.com/cypress-io/cypress/tree/develop/packages/errors) | `"rimraf": "3.0.2"`                                |
| [packages/runner](https://github.com/cypress-io/cypress/tree/develop/packages/runner) | `"rimraf": "^5.0.1"`                               |

- [rimraf](https://www.npmjs.com/package/rimraf) latest `5.x` is currently `5.0.10`
- [rimraf current](https://github.com/isaacs/rimraf/blob/main/package.json) `6.x` restricts to `"node": "20 || >=22"`

### PR DESCRIPTION
**Pull request by gweesin**
- Closes #30148 

### Steps to test

```shell
cd npm/puppeteer

# generate dist
yarn build

# remove dist and generate dist
yarn build
```

the same as `/errors/`, `/runner/` dir

### How has the user experience changed?

No change.

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by MikeMcC399** [comment]
@gweesin 

There are breaking changes for [rimraf](https://www.npmjs.com/package/rimraf) from v3 -> v4 -> v5 which need to be accounted for, see https://github.com/isaacs/rimraf/blob/v5.0.10/README.md

The following `build` command using `gulp` fails:

```shell
yarn workspace @packages/extension build
```

```text
@packages/extension: [00:00:59] Using gulpfile ~/cypress/packages/extension/gulpfile.ts
@packages/extension: [00:00:59] Starting 'build'...
@packages/extension: [00:00:59] Starting 'clean'...
@packages/extension: [00:00:59] 'clean' errored after 1.15 ms
@packages/extension: [00:00:59] Error: invalid rimraf options
@packages/extension:     at assertRimrafOptions (/root/cypress/node_modules/rimraf/src/opt-arg.ts:23:11)
@packages/extension:     at optArgT (/root/cypress/node_modules/rimraf/src/opt-arg.ts:54:22)
@packages/extension:     at optArg (/root/cypress/node_modules/rimraf/src/opt-arg.ts:82:57)
@packages/extension:     at /root/cypress/node_modules/rimraf/src/index.ts:30:27
@packages/extension:     at clean (/root/cypress/packages/extension/gulpfile.ts:16:9)
@packages/extension:     at bound (node:domain:433:15)
@packages/extension:     at runBound (node:domain:444:12)
@packages/extension:     at asyncRunner (/root/cypress/node_modules/async-done/index.js:55:18)
@packages/extension:     at processTicksAndRejections (node:internal/process/task_queues:77:11)
@packages/extension: [00:00:59] 'build' errored after 8.37 ms
```

So this means that the following would need to be changed:

https://github.com/cypress-io/cypress/blob/195cdb11a2af4b124bb15f8a38d62f058614a59d/packages/extension/gulpfile.ts#L15-L17

**Comment by jennifer-shehane** [comment]
Running the tests

**Comment by gweesin** [comment]
> Running the tests

I'm sorry my tests running blocked because of my Windows OS(#30157)

File: packages/extension/gulpfile.ts [hunk 1]
```diff
@@ -13,7 +13,7 @@ async function cypressIcons () {
 }
 
 const clean = (done) => {
-  rimraf('dist', done)
+  rimraf('dist').then(() => done())
 }
```
Comment (ryanthemanuel) [inline packages/extension/gulpfile.ts]: We could switch this to be even cleaner by making this an async function.

```javascript
const clean = async () => {
  await rimraf('dist')
}
```

**Comment by jennifer-shehane** [comment]
@gweesin Can you address the comment left to refactor the code here?

**Comment by gweesin** [comment]
> @gweesin Can you address the comment left to refactor the code here?

Of course. I never use gulp before, so I'm curious if it's possible to receive a Promise instead of callback function. I will try it later

**Comment by ryanthemanuel** [comment]
> > @gweesin Can you address the comment left to refactor the code here?
> 
> Of course. I never use gulp before, so I'm curious if it's possible to receive a Promise instead of callback function. I will try it later

@gweesin, Here's the relevant gulp documentation on async functions if you're curious to learn more about it: https://gulpjs.com/docs/en/getting-started/async-completion/#using-asyncawait

**Comment by gweesin** [comment]
@ryanthemanuel thank you, this is really helpful

File: packages/extension/gulpfile.ts (hunk 1 above)
Comment (gweesin) [reply packages/extension/gulpfile.ts]: I've found that in Gulp, returning a Promise is enough to make a task asynchronous. Therefore, returning the rimraf execution result directly is sufficient.

**Review by ryanthemanuel — APPROVED** [review]

**Comment by MikeMcC399** [comment]
@jennifer-shehane / @ryanthemanuel 

Should this be a `dependency:` commit instead of a `chore(deps):`?

Currently `rimraf@3.0.2` is packaged into the Cypress binary and this PR will bump it to `rimraf@5.0.10`.

**Comment by gweesin** [comment]
@MikeMcC399 Could I assume that devDependencies are not related to users, so they should be classified as `chore(deps)` rather than `dependency`?

**Comment by MikeMcC399** [comment]
@gweesin 

> Could I assume that devDependencies are not related to users, so they should be classified as `chore(deps)` rather than `dependency`?

Let me check. I may have made an incorrect assumption previously.

**Comment by MikeMcC399** [comment]
@gweesin 

My apologies! 🤭 I ran `yarn binary-build` on your branch and it left `rimraf` at version `3.0.2`, so your commit message is correct.
Please disregard my question about this and I will hide it as "resolved".

Edit: This is caused by

https://github.com/cypress-io/cypress/blob/b6b0b490d61b898283c8f6b31ef821c774c045ac/npm/webpack-dev-server/package.json#L28
as [webpack-dev-server@4.15.2](https://github.com/webpack/webpack-dev-server/blob/11bfcde1cbb291c349a61332937618685a98dd74/package.json#L69) has a dependency on `"rimraf": "^3.0.2"`

**Review by cacieprins — APPROVED** [review]

**Pull request merged**